Objects/Converters/ConverterRevit/ConverterRevitShared/Models/Element2DOutlineBuilder.cs

```diff
@@ -0,0 +1,254 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using ConverterRevitShared.Extensions;
+using Objects.Geometry;
+
+namespace ConverterRevitShared.Models
+{
+  /// <summary>
+  /// Responsible for taking the outermost loop of an analytical 2D element (as a list of points) and
+  /// adding back the sections that have been removed due to openings being present. This way, instead of sending
+  /// a wall with a door as 8-sided n-gon, we'll send it as a rectangle with the polyline representing the door
+  /// opening in a separate property. The latter format is much easier for structural software to represent and
+  /// is a more clear representation of the design intent.
+  /// </summary>
+  internal class Element2DOutlineBuilder
+  {
+    private const double POINT_TOLERANCE = .1;
+
+    private readonly List<Polyline> openingPolylines;
+    private readonly List<Point> outlinePoints;
+
+    public Element2DOutlineBuilder(List<Polyline> openingPolylines, List<Point> outlinePoints)
+    {
+      this.openingPolylines = openingPolylines;
+      this.outlinePoints = outlinePoints;
+    }
+
+    private void AddPointsToOutline(List<Point> pointsToAdd, int indexToBeInserted)
+    {
+      if (outlinePoints.Count == 0)
+      {
+        if (indexToBeInserted > 0)
+        {
+          throw new ArgumentException($"Outline currently has 0 points, cannot add point at index {indexToBeInserted}");
+        }
+        outlinePoints.AddRange(pointsToAdd);
+      }
+      else
+      {
+        AddPointsToPopulatedOutline(pointsToAdd, indexToBeInserted);
+      }
+    }
+
+    private void AddPointsToPopulatedOutline(List<Point> pointsToAdd, int? indexToBeInserted)
+    {
+      int insertIndex = indexToBeInserted == null ? outlinePoints.Count : indexToBeInserted.Value;
```

Comment:
If you need the nullable, I think indexToBeInserted.HasValue than the null check might be easier to read but I'm not precious about it

Reply:
I think this is still a recommendation but your choice